pkg/addon/utils.go

```diff
@@ -0,0 +1,201 @@
+limitations under the License.
+*/
+
+package addon
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+
+	errors "github.com/pkg/errors"
+
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	rest "k8s.io/client-go/rest"
+
+	"github.com/oam-dev/kubevela/apis/core.oam.dev/v1beta1"
+	"github.com/oam-dev/kubevela/pkg/definition"
+	"github.com/oam-dev/kubevela/pkg/oam"
+	"github.com/oam-dev/kubevela/pkg/oam/util"
+)
+
+const (
+	compDefAnnotation         = "addon.oam.dev/componentDefinitions"
+	traitDefAnnotation        = "addon.oam.dev/traitDefinitions"
+	workflowStepDefAnnotation = "addon.oam.dev/workflowStepDefinitions"
+	defKeytemplate            = "addon-%s-%s"
+)
+
+// parse addon's created x-defs in addon-app's annotation, this will be used to check whether app still using it while disabling.
+func passDefInAppAnnotation(defs []*unstructured.Unstructured, app *v1beta1.Application) error {
+	var comps, traits, workflowSteps []string
+	for _, def := range defs {
+		switch def.GetObjectKind().GroupVersionKind().Kind {
+		case v1beta1.ComponentDefinitionKind:
+			comps = append(comps, def.GetName())
+		case v1beta1.TraitDefinitionKind:
+			traits = append(traits, def.GetName())
+		case v1beta1.WorkflowStepDefinitionKind:
+			workflowSteps = append(workflowSteps, def.GetName())
+		default:
+			return fmt.Errorf("cannot handle definition types %s, name %s", def.GetObjectKind().GroupVersionKind().Kind, def.GetName())
+		}
+	}
+	if len(comps) != 0 {
+		c, _ := json.Marshal(comps)
+		app.SetAnnotations(util.MergeMapOverrideWithDst(app.GetAnnotations(), map[string]string{compDefAnnotation: string(c)}))
+	}
+	if len(traits) != 0 {
+		t, _ := json.Marshal(traits)
+		app.SetAnnotations(util.MergeMapOverrideWithDst(app.GetAnnotations(), map[string]string{traitDefAnnotation: string(t)}))
+	}
+	if len(workflowSteps) != 0 {
+		w, _ := json.Marshal(workflowSteps)
+		app.SetAnnotations(util.MergeMapOverrideWithDst(app.GetAnnotations(), map[string]string{workflowStepDefAnnotation: string(w)}))
+	}
+	return nil
+}
+
+// check whether this addon has been used by some applications
+func checkAddonHasBeenUsed(ctx context.Context, k8sClient client.Client, name string, addonApp v1beta1.Application, config *rest.Config) ([]v1beta1.Application, error) {
+	apps := v1beta1.ApplicationList{}
+	if err := k8sClient.List(ctx, &apps, client.InNamespace("")); err != nil {
+		return nil, err
+	}
+
+	if len(apps.Items) == 0 {
+		return nil, nil
+	}
+
+	createdDefs := make(map[string]bool)
+	for key, defNames := range addonApp.GetAnnotations() {
+		switch key {
+		case compDefAnnotation, traitDefAnnotation, workflowStepDefAnnotation:
+			if err := merge2DefMap(key, defNames, createdDefs); err != nil {
+				return nil, err
+			}
+		}
+	}
+
+	if len(createdDefs) == 0 {
+		if err := findLegacyAddonDefs(ctx, k8sClient, name, addonApp.GetLabels()[oam.LabelAddonRegistry], config, createdDefs); err != nil {
```

Comment:
This will make the uninstall process very slow, can we make it only done once like migration?

Reply:
good suggestion, fixed